### PR TITLE
add OpenRGB

### PR DESCRIPTION
**Pull request by 0xdevalias**
**Important:** *Do not tick a checkbox if you haven’t performed its action.* Honesty is indispensable for a smooth review process.

_In the following questions `<cask>` is the token of the cask you're submitting._

After making all changes to a cask, verify:

- [x] The submission is for [a stable version](https://docs.brew.sh/Acceptable-Casks#stable-versions) or [documented exception](https://docs.brew.sh/Acceptable-Casks#but-there-is-no-stable-version).
- [x] `brew audit --cask --online <cask>` is error-free.
- [x] `brew style --fix <cask>` reports no offenses.

Additionally, **if adding a new cask**:

- [x] Named the cask according to the [token reference](https://docs.brew.sh/Cask-Cookbook#token-reference).
- [x] Checked the cask was not [already refused](https://github.com/Homebrew/homebrew-cask/search?q=is%3Aclosed&type=Issues).
- [x] Checked the cask is submitted to [the correct repo](https://docs.brew.sh/Acceptable-Casks#finding-a-home-for-your-cask).
- [x] `brew audit --new-cask <cask>` worked successfully.
- [x] `brew install --cask <cask>` worked successfully.
- [x] `brew uninstall --cask <cask>` worked successfully.

### CODE REVIEW

**Comment by 0xdevalias** [comment]
Someone else originally created a `Cask` for this that was mentioned on this issue:

- https://gitlab.com/CalcProgrammer1/OpenRGB/-/issues/1824

Which had the following `Cask` source:

- https://github.com/muneebmahmed/homebrew-cask/blob/add-openrgb/Casks/openrgb.rb

---

The `Cask` in this PR was created from scratch by me, though I didn't include a `zap` clause. I haven't validated if the clause from the above older `Cask` is valid/complete, but it may be worth adding something like it to this PR potentially?

```ruby
  zap trash: [
    "~/Library/Saved Application State/com.yourcompany.OpenRGB.savedState",
    "~/.config/OpenRGB",
  ]
```

**Comment by 0xdevalias** [comment]
Also, I'm not really sure the best place/way to raise this (is an issue in this repo ok?), but if the `git` `livecheck` strategy supported passing the commit hash as well as the tag version, then the `livecheck` for this could have been much simpler/more robust/etc.

It also seems like the pattern I used here (or slight variations on it) could be used as a defactor `GitLab Releases` strategy, and could be useful to encode directly into homebrew as such to simplify cask definitions.

File: Casks/openrgb.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  version "0.8,fb88964"
+
+  on_intel do
+    sha256 "3d04b3caf0253e23a6fa7f854ee070cea7e589f3f1972d1975b512660d97fb1d"
+    url "https://openrgb.org/releases/release_#{version.before_comma}/OpenRGB_#{version.before_comma}_MacOS_Intel_#{version.after_comma}.zip"
+  end
+  on_arm do
+    sha256 "564a08138900d3e96bd045dba45bf3ac740720eb66d413cf4564f36218e177ea"
+    url "https://openrgb.org/releases/release_#{version.before_comma}/OpenRGB_#{version.before_comma}_MacOS_ARM64_#{version.after_comma}.zip"
+  end
```
Comment (SMillerDev) [inline Casks/openrgb.rb:11]:
```suggestion
  arch arm: "ARM64", intel: "Intel"
  sha256 arm: "564a08138900d3e96bd045dba45bf3ac740720eb66d413cf4564f36218e177ea",
                intel: "3d04b3caf0253e23a6fa7f854ee070cea7e589f3f1972d1975b512660d97fb1d"
    url "https://openrgb.org/releases/release_#{version.before_comma}/OpenRGB_#{version.before_comma}_MacOS_#{arch}_#{version.after_comma}.zip"
```

**Comment by SMillerDev** [comment]
> The Cask in this PR was created from scratch by me, though I didn't include a zap clause.

You should create one for this to be accepted. The [CreateZap](https://github.com/nrlquaker/homebrew-createzap) tool can help you with this.

> Also, I'm not really sure the best place/way to raise this (is an issue in this repo ok?), but if the git livecheck strategy supported passing the commit hash as well as the tag version, then the livecheck for this could have been much simpler/more robust/etc.

That would have to be in Homebrew/brew. Git tags by definition only point to a single commit hash though so that information is redundant, unless you retag. 
The [git manual](https://git-scm.com/docs/git-tag#_on_re_tagging) says re-tagging is "the insane thing" to do though.

**Comment by 0xdevalias** [comment]
> That would have to be in Homebrew/brew. Git tags by definition only point to a single commit hash though so that information is redundant, unless you retag.
> The [git manual](https://git-scm.com/docs/git-tag#_on_re_tagging) says re-tagging is "the insane thing" to do though.

It's more that to construct the download link we need both the git commit hash, and the version; and based on my naive understanding of how to achieve that in a `livecheck` block, it's not currently possible to achieve that with the git strategy; yet the rough proposal above would allow it to be so.

**Comment by 0xdevalias** [comment]
> > The Cask in this PR was created from scratch by me, though I didn't include a zap clause.
> 
> You should create one for this to be accepted. The [CreateZap](https://github.com/nrlquaker/homebrew-createzap) tool can help you with this.

Using that gives me:

```ruby
zap trash: [
  "~/Library/Preferences/com.yourcompany.OpenRGB.plist",
  "~/Library/Saved Application State/com.yourcompany.OpenRGB.savedState",
]
```

Yet when I ran the app it also created stuff in `~/.config/OpenRGB`, so the full zap clause should probably be:

```ruby
zap trash: [
  "~/Library/Preferences/com.yourcompany.OpenRGB.plist",
  "~/Library/Saved Application State/com.yourcompany.OpenRGB.savedState",
  "~/.config/OpenRGB"
]
```

File: Casks/openrgb.rb (hunk 1 above)
Comment (0xdevalias) [reply Casks/openrgb.rb]: [`0be662b` (#138933)](https://github.com/Homebrew/homebrew-cask/pull/138933/commits/0be662bb58b662aca8bc1a4c73b05d57bee3f15a)

**Comment by samford** [comment]
> Also, I'm not really sure the best place/way to raise this (is an issue in this repo ok?), but if the `git` `livecheck` strategy supported passing the commit hash as well as the tag version, then the `livecheck` for this could have been much simpler/more robust/etc.

We've discussed surfacing the Git tag hash from livecheck in a different context (i.e., it's needed to be able to programmatically version bump formulae that use `tag`/`revision` for the `stable` URL) but it wouldn't be too hard to optionally pass the hash to a `strategy` block. That said, it wouldn't be a solution in this particular instance because this cask is using a release asset, so we want livecheck to report a new version when it's _released_ and not just tagged (sometimes there's a notable gap between these actions).

> It also seems like the pattern I used here (or slight variations on it) could be used as a defactor `GitLab Releases` strategy, and could be useful to encode directly into homebrew as such to simplify cask definitions.

At the moment, I believe the only other cask using a similar approach is `edfbrowser`. I wouldn't be averse to the idea of creating a dedicated strategy if this approach becomes more commonplace but it would be premature based on current usage.

> It also wasn't clear to me where I could even find documentation on all of the supported livecheck strategies

The main livecheck documentation can be found on the related [docs.brew.sh page](https://docs.brew.sh/Brew-Livecheck) but the strategy-specific information can be found in the `Homebrew::Livecheck::Strategy` module on [rubydoc.brew.sh](https://rubydoc.brew.sh/Homebrew/Livecheck/Strategy.html). I need to update the main documentation a bit and I plan on expanding strategy-specific documentation in the future but that's most of the documentation outside of reading [the code in Homebrew/brew](https://github.com/Homebrew/brew/tree/master/Library/Homebrew/livecheck).

-----

That said, this `livecheck` block should be checking the same source as the download `url` (not Gitlab, which is separate). I'll add a suggested change in a moment.

File: Casks/openrgb.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+         intel: "3d04b3caf0253e23a6fa7f854ee070cea7e589f3f1972d1975b512660d97fb1d"
+
+  url "https://openrgb.org/releases/release_#{version.csv.first}/OpenRGB_#{version.csv.first}_MacOS_#{arch}_#{version.csv.second}.zip"
+  name "OpenRGB"
+  desc "Open source RGB lighting control that doesn't depend on manufacturer software"
+  homepage "https://openrgb.org/"
+
+  livecheck do
+    url "https://gitlab.com/CalcProgrammer1/OpenRGB/-/releases.json"
+    regex(%r{/releases/release_([^/]+)/OpenRGB_\1+_MacOS_Intel_([a-f0-9]+).zip}i)
```
Comment (samford) [inline Casks/openrgb.rb:15]:
```suggestion
    url "https://openrgb.org/releases.html"
    regex(/href=.*?OpenRGB[._-]v?(\d+(?:\.\d+)+)[._-]MacOS[._-]#{arch}[._-](\h+)\.zip/i)
```

This aligns the check with the source of the cask `url`, which is preferable in this context.

**Review by miccal — APPROVED** [review]

**Comment by 0xdevalias** [comment]
> At the moment, I believe the only other cask using a similar approach is `edfbrowser`. I wouldn't be averse to the idea of creating a dedicated strategy if this approach becomes more commonplace but it would be premature based on current usage.

@samford I believe there were a couple of variations I saw for GitLab, I just ended up choosing this one as the 'better' looking of them:

- https://github.com/search?q=repo%3AHomebrew%2Fhomebrew-cask%20livecheck%20gitlab&type=code

## Reading the releases page by HTML

IMO worse than reading the releases via JSON/API.

https://github.com/Homebrew/homebrew-cask/blob/ee5d8242c56260eae340c78b8078efccc81a702c/Casks/librewolf.rb#L19-L27
https://github.com/Homebrew/homebrew-cask/blob/ee5d8242c56260eae340c78b8078efccc81a702c/Casks/cura-lulzbot.rb#L11-L17

## Reading the GitLab releases by API/JSON

IMO better than reading the release page HTML.

https://github.com/Homebrew/homebrew-cask/blob/ee5d8242c56260eae340c78b8078efccc81a702c/Casks/edfbrowser.rb#L11-L18

Using the git strategy:
https://github.com/Homebrew/homebrew-cask/blob/ee5d8242c56260eae340c78b8078efccc81a702c/Casks/studiolinkstandalone.rb#L21-L25

---

At the very least, if not adding a new custom strategy, it might be useful to add a linting rule to suggest/standardise on using either the releases page HTML, or the releases API/JSON for consistency (and so others don't need to guess which is best to use in future)

**Comment by 0xdevalias** [comment]
> You should create one for this to be accepted. The [CreateZap](https://github.com/nrlquaker/homebrew-createzap) tool can help you with this.

@SMillerDev It would be nice if that was included as part of the template when using `brew create --cask` so that all of the 'required knowledge' is encoded/as close to the source as possible; rather than having to rely on individual contributors having all the knowledge/up to date practices in their heads (particularly given the contributing docs seem a bit all over the place)

**Pull request merged**